File: src/main/java/com/fengsheng/handler/end_main_phase_tos.java
package com.fengsheng.handler;

import com.fengsheng.HumanPlayer;
import com.fengsheng.phase.MainPhaseIdle;
import com.fengsheng.phase.SendPhaseStart;
import com.fengsheng.protos.Fengsheng;
import org.apache.log4j.Logger;

public class end_main_phase_tos extends AbstractProtoHandler<Fengsheng.end_main_phase_tos> {
    private static final Logger log = Logger.getLogger(end_main_phase_tos.class);

    @Override
    protected void handle0(HumanPlayer r, Fengsheng.end_main_phase_tos pb) {
        if (!r.checkSeq(pb.getSeq())) {
            log.error("操作太晚了, required Seq: " + r.getSeq() + ", actual Seq: " + pb.getSeq());
            return;
        }
        if (!(r.getGame().getFsm() instanceof MainPhaseIdle fsm) || r != fsm.player()) {
            log.error("不是你的回合的出牌阶段");
            return;
        }
        r.incrSeq();
        r.getGame().resolve(new SendPhaseStart(fsm.player()));
    }
}

File: src/main/kotlin/handler/end_main_phase_tos.kt
package com.fengsheng.handler

import com.fengsheng.HumanPlayer
import com.fengsheng.phase.MainPhaseIdle
import com.fengsheng.phase.SendPhaseStart
import com.fengsheng.protos.Fengsheng
import com.fengsheng.skill.MainPhaseSkill
import org.apache.log4j.Logger

class end_main_phase_tos : AbstractProtoHandler<Fengsheng.end_main_phase_tos>() {
    override fun handle0(r: HumanPlayer, pb: Fengsheng.end_main_phase_tos) {
        if (!r.checkSeq(pb.seq)) {
            log.error("操作太晚了, required Seq: ${r.seq}, actual Seq: ${pb.seq}")
            r.sendErrorMessage("操作太晚了")
            return
        }
        val fsm = r.game!!.fsm as? MainPhaseIdle
        if (r !== fsm?.player) {
            log.error("不是你的回合的出牌阶段")
            r.sendErrorMessage("不是你的回合的出牌阶段")
            return
        }
        if (!r.game!!.mainPhaseAlreadyNotify && r.skills.any { it is MainPhaseSkill && it.mainPhaseNeedNotify(r) }) {
            r.game!!.mainPhaseAlreadyNotify = true
            r.sendErrorMessage("还有未发动的技能，真的要结束出牌阶段吗？")
            return
        }
        r.incrSeq()
        r.game!!.resolve(SendPhaseStart(fsm.player))
    }

    companion object {
        private val log = Logger.getLogger(end_main_phase_tos::class.java)
    }
}